### PR TITLE
[Datasets] Shuffled data loading support

### PR DESCRIPTION
**Pull request by clarkzinzow**
A few hotfixes for shuffled data loading support.

1. Move the block refs when shuffling and delete intermediate block refs in order to reduce the number of dataset copies during shuffling.
2. Pass remote args during reads as submission-time task options, allowing for more options (e.g. placement groups).
3. Add round-robin resource-based load balancing (private, undocumented) option to reading and shuffling via a few environment variables.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/impl/shuffle.py [hunk 1]
```diff
@@ -13,23 +14,41 @@
...
-                   *,
-                   random_shuffle: bool = False,
-                   random_seed: Optional[int] = None) -> BlockList[T]:
+def simple_shuffle(
+        input_blocks: BlockList[T],
+        output_num_blocks: int,
+        *,
+        random_shuffle: bool = False,
+        random_seed: Optional[int] = None,
+        map_ray_remote_args: Dict[str, Any] = None,
```
Comment (scv119) [inline python/ray/data/impl/shuffle.py:23]: should this be typed `Optional[Dict[str, Any]]` ?

File: python/ray/data/dataset.py [hunk 2]
```diff
@@ -349,14 +350,33 @@ def random_shuffle(self,
...
+            self.move_blocks(),
+            num_blocks or num_blocks,
             random_shuffle=True,
-            random_seed=seed)
+            random_seed=seed,
+            _round_robin_resource_provider=round_robin_resource_provider)
         return Dataset(new_blocks)
 
+    def move_blocks(self):
+        blocks = self._blocks
```
Comment (ericl) [inline python/ray/data/dataset.py:376]: Make this private. Also, can we raise a good error message if self.blocks is None and we try to use the dataset?

File: python/ray/data/impl/shuffle.py [hunk 3]
```diff
@@ -44,10 +63,13 @@ def simple_shuffle(input_blocks: BlockList[T],
...
-        shuffle_reduce.remote(
-            *[shuffle_map_out[i][j] for i in range(input_num_blocks)])
+        shuffle_reduce.options(
+            **reduce_ray_remote_args,
+            resources=next(reduce_resource_iter)).remote(
+                *[shuffle_map_out[i][j] for i in range(input_num_blocks)])
         for j in range(output_num_blocks)
     ]
+    del shuffle_map_out
     new_blocks, new_metadata = zip(*shuffle_reduce_out)
```
Comment (ericl) [inline python/ray/data/impl/shuffle.py:73]: Comment for why

File: python/ray/data/impl/shuffle.py [hunk 4]
```diff
@@ -13,23 +14,41 @@
...
 
     shuffle_map_out = [
-        shuffle_map.remote(block, i, output_num_blocks, random_shuffle,
-                           random_seed) for i, block in enumerate(input_blocks)
+        shuffle_map.options(
+            **map_ray_remote_args, resources=next(map_resource_iter)).remote(
+                block, i, output_num_blocks, random_shuffle, random_seed)
+        for i, block in enumerate(input_blocks)
     ]
+    del input_blocks
```
Comment (ericl) [inline python/ray/data/impl/shuffle.py:51]: Comment for del

File: python/ray/data/dataset.py [hunk 5]
```diff
@@ -349,14 +350,33 @@ def random_shuffle(self,
         Returns:
             The shuffled dataset.
         """
+        # Check for spread resource labels in environment variable, and use
+        # the given labels for round-robin node-based scheduling.
+        shuffle_spread_custom_resource_labels = os.getenv(
+            "RAY_DATASETS_SHUFFLE_SPREAD_CUSTOM_RESOURCE_LABELS", None)
```
Comment (ericl) [inline python/ray/data/dataset.py:356]: Move this code into simple shuffle impl?

File: python/ray/data/dataset.py [hunk 6]
```diff
@@ -326,7 +326,8 @@ def repartition(self, num_blocks: int) -> "Dataset[T]":
     def random_shuffle(self,
                        *,
                        seed: Optional[int] = None,
-                       num_blocks: Optional[int] = None) -> "Dataset[T]":
+                       num_blocks: Optional[int] = None,
+                       move: Optional[bool] = False) -> "Dataset[T]":
```
Comment (ericl) [inline python/ray/data/dataset.py:330]:
```suggestion
                       _move: Optional[bool] = False) -> "Dataset[T]":
```

File: python/ray/data/dataset.py [hunk 7]
```diff
@@ -345,18 +346,28 @@ def random_shuffle(self,
                 based on system randomness.
             num_blocks: The number of output blocks after the shuffle, or None
                 to retain the number of blocks.
+            move: Whether the blocks for the source dataset should be moved
```
Comment (ericl) [inline python/ray/data/dataset.py:349]: Let's keep this undocumented for now. We should re-think lazy vs eager execution management and come up with a public API holistically.

File: python/ray/data/impl/block_list.py [hunk 8]
```diff
@@ -20,7 +20,20 @@ def set_metadata(self, i: int, metadata: BlockMetadata) -> None:
     def get_metadata(self) -> List[BlockMetadata]:
         return self._metadata.copy()
 
+    def copy(self) -> "BlockList":
+        return BlockList(self._blocks, self._metadata)
+
+    def clear(self):
+        self._blocks = None
+
+    def _check_if_cleared(self):
```
Comment (ericl) [inline python/ray/data/impl/block_list.py:29]: :+1:

**Review by ericl — APPROVED** [review]

**Review by scv119 — APPROVED** [review]

**Comment by scv119** [comment]
LGTM!

**Comment by scv119** [comment]
linter failure looks relevant.

**Comment by scv119** [comment]
huh I was debugging why this work with dataset, but it doesn't work with dataset_pipeline. I realized this is because the ENV is not propagated to worker/remote node, and in pipelined dataset we only set ENV in driver.

**Pull request merged**